kong/plugins/key-auth/handler.lua

```diff
@@ -41,23 +41,22 @@ end
 local function set_consumer(consumer, credential)
   local const = constants.HEADERS
 
-  local new_headers = {
-    [const.CONSUMER_ID]        = consumer.id,
-    [const.CONSUMER_CUSTOM_ID] = tostring(consumer.custom_id),
-    [const.CONSUMER_USERNAME]  = consumer.username,
-  }
-
   kong.client.authenticate(consumer, credential)
 
+  kong.service.request.set_header(const.CONSUMER_ID, consumer.id)
+  kong.service.request.set_header(const.CONSUMER_CUSTOM_ID, tostring(consumer.custom_id))
+  kong.service.request.set_header(const.CONSUMER_USERNAME, consumer.username)
```

Comment:
`username is optional`, you will need a guard here.

Reply:
There is no test with `consumer` with username, and the `tostring` there is I think wong, it should be something like:

```lua
if consumer.custom_id then
  kong.service.request.set_header(const.CONSUMER_CUSTOM_ID, consumer.custom_id)
else
  kong.service.request.clear_header(const.CONSUMER_CUSTOM_ID, consumer.custom_id)
end
```

The issue is that `set_header` in `pdk` does not allow `nil`, as did the `OpenResty`'s one. I am also making changes here:
https://github.com/Kong/kong/pull/3976

I can take your changes and port them there too, is that ok?

Reply:
@bungle yeah that’s be great, thanks. It looks like your work is already covering this anyhow. Thanks!